File: Src/java/cql-to-elm-cli/src/main/java/org/cqframework/cql/cql2elm/cli/Main.java
package org.cqframework.cql.cql2elm.cli;

import static java.nio.file.FileVisitResult.CONTINUE;
import static org.cqframework.cql.cql2elm.CqlTranslator.fromFile;

import java.io.File;
import java.io.IOException;
import java.io.PrintWriter;
import java.nio.file.FileVisitResult;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.SimpleFileVisitor;
import java.nio.file.attribute.BasicFileAttributes;
import java.util.HashMap;
import java.util.Map;
import joptsimple.OptionParser;
import joptsimple.OptionSet;
import joptsimple.OptionSpec;
import org.cqframework.cql.cql2elm.*;
import org.cqframework.cql.cql2elm.quick.FhirLibrarySourceProvider;
import org.cqframework.cql.elm.tracking.TrackBack;
import org.cqframework.fhir.npm.LibraryLoader;
import org.cqframework.fhir.npm.NpmLibrarySourceProvider;
import org.cqframework.fhir.npm.NpmPackageManager;
import org.cqframework.fhir.utilities.IGContext;
import org.hl7.cql.model.ModelIdentifier;
import org.hl7.cql.model.ModelInfoProvider;
import org.hl7.cql.model.NamespaceInfo;
import org.hl7.cql.model.NamespaceManager;
import org.hl7.elm_modelinfo.r1.ModelInfo;
import org.hl7.elm_modelinfo.r1.serializing.ModelInfoReaderFactory;

public class Main {
    public static ModelInfoProvider getModelInfoProvider(File modelInfoXML) {
        try {
            final ModelInfo modelInfo =
                    ModelInfoReaderFactory.getReader("application/xml").read(modelInfoXML);
            return (ModelIdentifier modelIdentifier) -> modelInfo;
        } catch (IOException e) {
            System.err.printf("Could not load model-info XML: %s%n", modelInfoXML);
            e.printStackTrace();
            System.exit(-1);
        }
        return null;
    }

    private static void outputExceptions(Iterable<CqlCompilerException> exceptions) {
        for (CqlCompilerException error : exceptions) {
            TrackBack tb = error.getLocator();
            String lines = tb == null
                    ? "[n/a]"
                    : String.format(
                            "[%d:%d, %d:%d]", tb.getStartLine(), tb.getStartChar(), tb.getEndLine(), tb.getEndChar());
            System.err.printf("%s:%s %s%n", error.getSeverity(), lines, error.getMessage());
        }
    }

    private static void writeELM(
            Path inPath,
            Path outPath,
            org.cqframework.cql.cql2elm.CqlTranslator.Format format,
            ModelInfoProvider modelProvider,
            IGContext igContext,
            CqlCompilerOptions options)
            throws IOException {

        System.err.println("================================================================================");
        System.err.printf("TRANSLATE %s%n", inPath);

        ModelManager modelManager;
        if (options.getOptions().contains(CqlCompilerOptions.Options.DisableDefaultModelInfoLoad)) {
            modelManager = new ModelManager(false);
        } else {
            modelManager = new ModelManager();
        }

        if (modelProvider != null) {
            modelManager.getModelInfoLoader().registerModelInfoProvider(modelProvider);
        }

        LibraryManager libraryManager = new LibraryManager(modelManager, options);
        modelManager
                .getModelInfoLoader()
                .registerModelInfoProvider(new DefaultModelInfoProvider(inPath.getParent()), true);
        libraryManager.getLibrarySourceLoader().registerProvider(new DefaultLibrarySourceProvider(inPath.getParent()));
        libraryManager.getLibrarySourceLoader().registerProvider(new FhirLibrarySourceProvider());

        NamespaceManager namespaceManager = libraryManager.getNamespaceManager();
        NamespaceInfo namespaceInfo = null;
        if (igContext != null) {
            NpmPackageManager pm = new NpmPackageManager(igContext.getSourceIg());
            pm.getNpmList().forEach(npm -> {
                NamespaceInfo newNamespace = new NamespaceInfo(npm.id(), npm.canonical());
                namespaceManager.ensureNamespaceRegistered(newNamespace);
            });
            LibraryLoader reader = new LibraryLoader(igContext.getFhirVersion());
            NpmLibrarySourceProvider sp = new NpmLibrarySourceProvider(pm.getNpmList(), reader, pm);
            libraryManager.getLibrarySourceLoader().registerProvider(sp);

            String packageId = igContext.getPackageId();
            String canonicalBase = igContext.getCanonicalBase();

            if (packageId != null && !packageId.isEmpty() && canonicalBase != null && !canonicalBase.isEmpty()) {
                namespaceInfo = new NamespaceInfo(packageId, canonicalBase);
            }
        }

        CqlTranslator translator = fromFile(namespaceInfo, inPath.toFile(), libraryManager);
        libraryManager.getLibrarySourceLoader().clearProviders();

        if (!translator.getErrors().isEmpty()) {
            System.err.println("Translation failed due to errors:");
            outputExceptions(translator.getExceptions());
        } else if (!options.getVerifyOnly()) {
            if (translator.getExceptions().isEmpty()) {
                System.err.println("Translation completed successfully.");
            } else {
                System.err.println("Translation completed with messages:");
                outputExceptions(translator.getExceptions());
            }
            try (PrintWriter pw = new PrintWriter(outPath.toFile(), "UTF-8")) {
                switch (format) {
                    case COFFEE:
                        pw.print("module.exports = ");
                        pw.println(translator.toJson());
                        break;
                    case JSON:
                        pw.println(translator.toJson());
                        break;
                    case XML:
                    default:
                        pw.println(translator.toXml());
                }
                pw.println();
            }
            System.err.println(String.format("ELM output written to: %s", outPath.toString()));
        }

        System.err.println();
    }

    @SuppressWarnings({"unchecked", "rawtypes"})
    public static void main(String[] args) throws IOException, InterruptedException {
        OptionParser parser = new OptionParser();
        OptionSpec<File> input = parser.accepts("input")
                .withRequiredArg()
                .ofType(File.class)
                .required()
                .describedAs(
                        "The name of the input file or directory. If a directory is given, all files ending in .cql will be processed");
        OptionSpec<File> model = parser.accepts("model")
                .withRequiredArg()
                .ofType(File.class)
                .describedAs(
                        "The name of an input file containing the model info to use for translation. Model info can also be provided through an implementation of ModelInfoProvider");
        OptionSpec<File> output = parser.accepts("output")
                .withRequiredArg()
                .ofType(File.class)
                .describedAs(
                        "The name of the output file or directory. If no output is given, an output file name is constructed based on the input name and target format");
        OptionSpec<org.cqframework.cql.cql2elm.CqlTranslator.Format> format = parser.accepts("format")
                .withRequiredArg()
                .ofType(org.cqframework.cql.cql2elm.CqlTranslator.Format.class)
                .defaultsTo(org.cqframework.cql.cql2elm.CqlTranslator.Format.XML)
                .describedAs("The target format for the output");
        OptionSpec<File> rootDir = parser.accepts("root-dir")
                .withOptionalArg()
                .ofType(File.class)
                .describedAs("Root directory of a FHIR IG project, used to resolve CQL namespaces");
        OptionSpec disableDefaultModelInfoLoad = parser.accepts("disable-default-modelinfo-load");
        OptionSpec verify = parser.accepts("verify");
        OptionSpec optimization = parser.accepts("date-range-optimization");
        OptionSpec annotations = parser.accepts("annotations");
        OptionSpec locators = parser.accepts("locators");
        OptionSpec resultTypes = parser.accepts("result-types");
        OptionSpec detailedErrors = parser.accepts("detailed-errors");
        OptionSpec errorLevel = parser.accepts("error-level")
                .withRequiredArg()
                .ofType(CqlCompilerException.ErrorSeverity.class)
                .defaultsTo(CqlCompilerException.ErrorSeverity.Info)
                .describedAs(
                        "Indicates the minimum severity message that will be reported. If no error-level is specified, all messages will be output");
        OptionSpec disableListTraversal = parser.accepts("disable-list-traversal");
        OptionSpec disableListDemotion = parser.accepts("disable-list-demotion");
        OptionSpec disableListPromotion = parser.accepts("disable-list-promotion");
        OptionSpec enableIntervalDemotion = parser.accepts("enable-interval-demotion");
        OptionSpec enableIntervalPromotion = parser.accepts("enable-interval-promotion");
        OptionSpec disableMethodInvocation = parser.accepts("disable-method-invocation");
        OptionSpec requireFromKeyword = parser.accepts("require-from-keyword");
        OptionSpec strict = parser.accepts("strict");
        OptionSpec debug = parser.accepts("debug");
        OptionSpec validateUnits = parser.accepts("validate-units");
        OptionSpec<LibraryBuilder.SignatureLevel> signatures = parser.accepts("signatures")
                .withRequiredArg()
                .ofType(LibraryBuilder.SignatureLevel.class)
                .defaultsTo(LibraryBuilder.SignatureLevel.None)
                .describedAs(
                        "Indicates whether signatures should be included for invocations in the output ELM. Differing will include invocation signatures that differ from the declared signature. Overloads will include declaration signatures when the operator or function has more than one overload with the same number of arguments as the invocation");
        OptionSpec<String> compatibilityLevel = parser.accepts("compatibility-level")
                .withRequiredArg()
                .ofType(String.class)
                .describedAs("Compatibility level for the translator, valid values are 1.3, 1.4, and 1.5");

        OptionSet options = parser.parse(args);

        final Path source = input.value(options).toPath();
        final Path destination = output.value(options) != null
                ? output.value(options).toPath()
                : source.toFile().isDirectory() ? source : source.getParent();
        final org.cqframework.cql.cql2elm.CqlTranslator.Format outputFormat = format.value(options);
        final LibraryBuilder.SignatureLevel signatureLevel = signatures.value(options);

        Map<Path, Path> inOutMap = new HashMap<>();
        if (source.toFile().isDirectory()) {
            if (destination.toFile().exists() && !destination.toFile().isDirectory()) {
                throw new IllegalArgumentException("Output must be a valid folder if input is a folder!");
            }

            Files.walkFileTree(source, new SimpleFileVisitor<Path>() {
                @Override
                public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
                    if (file.toFile().getName().endsWith(".cql")
                            || file.toFile().getName().endsWith(".CQL")) {
                        Path destinationFolder = destination.resolve(source.relativize(file.getParent()));
                        if (!destinationFolder.toFile().exists()
                                && !destinationFolder.toFile().mkdirs()) {
                            System.err.printf("Problem creating %s%n", destinationFolder);
                        }
                        inOutMap.put(file, destinationFolder);
                    }
                    return CONTINUE;
                }
            });
        } else {
            inOutMap.put(source, destination);
        }

        for (Map.Entry<Path, Path> inOut : inOutMap.entrySet()) {
            Path in = inOut.getKey();
            Path out = inOut.getValue();
            if (out.toFile().isDirectory()) {
                // Use input filename with ".xml", ".json", or ".coffee" extension
                String name = in.toFile().getName();
                if (name.lastIndexOf('.') != -1) {
                    name = name.substring(0, name.lastIndexOf('.'));
                }
                switch (outputFormat) {
                    case JSON:
                        name += ".json";
                        break;
                    case COFFEE:
                        name += ".coffee";
                        break;
                    case XML:
                    default:
                        name += ".xml";
                        break;
                }
                out = out.resolve(name);
            }

            if (out.equals(in)) {
                throw new IllegalArgumentException("input and output file must be different!");
            }

            IGContext igContext = null;
            if (options.has(rootDir)) {
                igContext = new IGContext();
                igContext.initializeFromIni(options.valueOf("root-dir") + File.separator + "ig.ini");
            }

            ModelInfoProvider modelProvider = null;
            if (options.has(model)) {
                final File modelFile = options.valueOf(model);
                if (!modelFile.exists() || modelFile.isDirectory()) {
                    throw new IllegalArgumentException("model must be a valid file!");
                }
                modelProvider = getModelInfoProvider(modelFile);
            }

            writeELM(
                    in,
                    out,
                    outputFormat,
                    modelProvider,
                    igContext,
                    new CqlCompilerOptions(
                            options.has(optimization),
                            options.has(debug) || options.has(annotations),
                            options.has(debug) || options.has(locators),
                            options.has(debug) || options.has(resultTypes),
                            options.has(verify),
                            options.has(detailedErrors), // Didn't include in debug, maybe should...
                            options.has(errorLevel)
                                    ? (CqlCompilerException.ErrorSeverity) options.valueOf(errorLevel)
                                    : CqlCompilerException.ErrorSeverity.Info,
                            options.has(strict) || options.has(disableListTraversal),
                            options.has(strict) || options.has(disableListDemotion),
                            options.has(strict) || options.has(disableListPromotion),
                            options.has(enableIntervalDemotion),
                            options.has(enableIntervalPromotion),
                            options.has(strict) || options.has(disableMethodInvocation),
                            options.has(requireFromKeyword),
                            options.has(validateUnits),
                            options.has(disableDefaultModelInfoLoad),
                            signatureLevel,
                            options.has(compatibilityLevel) ? options.valueOf(compatibilityLevel) : null));
        }
    }
}

File: Src/java/cql-to-elm-cli/src/main/kotlin/org/cqframework/cql/cql2elm/cli/Main.kt
package org.cqframework.cql.cql2elm.cli

import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.io.PrintWriter
import java.nio.file.FileVisitResult
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.SimpleFileVisitor
import java.nio.file.attribute.BasicFileAttributes
import joptsimple.OptionParser
import joptsimple.OptionSpec
import kotlin.system.exitProcess
import kotlinx.io.asSource
import kotlinx.io.buffered
import org.cqframework.cql.cql2elm.CqlCompilerException
import org.cqframework.cql.cql2elm.CqlCompilerOptions
import org.cqframework.cql.cql2elm.CqlTranslator
import org.cqframework.cql.cql2elm.CqlTranslator.Companion.fromFile
import org.cqframework.cql.cql2elm.DefaultLibrarySourceProvider
import org.cqframework.cql.cql2elm.DefaultModelInfoProvider
import org.cqframework.cql.cql2elm.LibraryBuilder.SignatureLevel
import org.cqframework.cql.cql2elm.LibraryManager
import org.cqframework.cql.cql2elm.ModelManager
import org.cqframework.cql.cql2elm.quick.FhirLibrarySourceProvider
import org.cqframework.fhir.npm.LibraryLoader
import org.cqframework.fhir.npm.NpmLibrarySourceProvider
import org.cqframework.fhir.npm.NpmModelInfoProvider
import org.cqframework.fhir.npm.NpmPackageManager
import org.cqframework.fhir.utilities.IGContext
import org.hl7.cql.model.ModelIdentifier
import org.hl7.cql.model.ModelInfoProvider
import org.hl7.cql.model.NamespaceInfo
import org.hl7.elm_modelinfo.r1.ModelInfo
import org.hl7.elm_modelinfo.r1.serializing.parseModelInfoXml
import org.hl7.fhir.utilities.npm.NpmPackage

object Main {
    @Suppress("PrintStackTrace")
    fun getModelInfoProvider(modelInfoXML: File): ModelInfoProvider {
        try {
            val source = FileInputStream(modelInfoXML).asSource().buffered()
            val modelInfo: ModelInfo = parseModelInfoXml(source)
            return { _: ModelIdentifier -> modelInfo } as ModelInfoProvider
        } catch (e: IOException) {
            System.err.printf("Could not load model-info XML: %s%n", modelInfoXML)
            e.printStackTrace()
            exitProcess(-1)
        }
    }

    private fun outputExceptions(exceptions: Iterable<CqlCompilerException>) {
        for (error in exceptions) {
            val tb = error.locator
            val lines =
                if (tb == null) "[n/a]"
                else ("[${tb.startLine}:${tb.startChar}, ${tb.endLine}:${tb.endChar}]")
            System.err.printf("%s:%s %s%n", error.severity, lines, error.message)
        }
    }

    @Suppress("LongParameterList", "LongMethod", "CyclomaticComplexMethod")
    @Throws(IOException::class)
    private fun writeELM(
        inPath: Path,
        outPath: Path,
        format: CqlTranslator.Format,
        modelProvider: ModelInfoProvider?,
        igContext: IGContext?,
        options: CqlCompilerOptions,
    ) {
        System.err.println(
            "================================================================================"
        )
        System.err.printf("TRANSLATE %s%n", inPath)
        val modelManager =
            if (options.options.contains(CqlCompilerOptions.Options.DisableDefaultModelInfoLoad)) {
                ModelManager(false)
            } else {
                ModelManager()
            }

        if (modelProvider != null) {
            modelManager.modelInfoLoader.registerModelInfoProvider(modelProvider)
        }

        val libraryManager = LibraryManager(modelManager, options)
        val parent = kotlinx.io.files.Path(inPath.parent.toString())
        modelManager.modelInfoLoader.registerModelInfoProvider(
            DefaultModelInfoProvider(parent),
            true,
        )
        libraryManager.librarySourceLoader.registerProvider(DefaultLibrarySourceProvider(parent))
        libraryManager.librarySourceLoader.registerProvider(FhirLibrarySourceProvider())

        val namespaceManager = libraryManager.namespaceManager
        var namespaceInfo: NamespaceInfo? = null
        if (igContext != null) {
            val pm = NpmPackageManager(igContext.sourceIg!!)
            @Suppress("ForbiddenComment")
            pm.npmList.forEach { npm: NpmPackage? ->
                val newNamespace = NamespaceInfo(npm!!.id(), npm.canonical())
                if (namespaceManager.resolveNamespaceUri(newNamespace.name) != null) {
                    // TODO: Logger.skip loading
                } else if (namespaceManager.getNamespaceInfoFromUri(newNamespace.uri) != null) {
                    // TODO: Logger.skip loading
                } else {
                    namespaceManager.ensureNamespaceRegistered(newNamespace)
                }
            }
            val reader = LibraryLoader(igContext.fhirVersion)
            val sp = NpmLibrarySourceProvider(pm.npmList, reader, pm)
            libraryManager.librarySourceLoader.registerProvider(sp)

            val mp = NpmModelInfoProvider(pm.npmList, reader, pm)
            modelManager.modelInfoLoader.registerModelInfoProvider(mp)

            val packageId = igContext.packageId
            val canonicalBase = igContext.canonicalBase

            if (!packageId.isNullOrBlank() && !canonicalBase.isNullOrBlank()) {
                namespaceInfo = NamespaceInfo(packageId, canonicalBase)
            }
        }

        val translator: CqlTranslator = fromFile(namespaceInfo, inPath.toString(), libraryManager)
        libraryManager.librarySourceLoader.clearProviders()

        if (!translator.errors.isEmpty()) {
            System.err.println("Translation failed due to errors:")
            outputExceptions(translator.exceptions)
        } else if (!options.verifyOnly) {
            if (translator.exceptions.isEmpty()) {
                System.err.println("Translation completed successfully.")
            } else {
                System.err.println("Translation completed with messages:")
                outputExceptions(translator.exceptions)
            }

            val os = FileOutputStream(outPath.toFile())
            PrintWriter(os).use { pw ->
                when (format) {
                    CqlTranslator.Format.COFFEE -> {
                        pw.print("module.exports = ")
                        pw.println(translator.toJson())
                    }

                    CqlTranslator.Format.JSON -> pw.println(translator.toJson())
                    CqlTranslator.Format.XML -> pw.println(translator.toXml())
                }
                pw.println()
            }
            System.err.println("ELM output written to: $outPath")
        }

        System.err.println()
    }

    @Suppress("LongMethod", "CyclomaticComplexMethod", "MemberNameEqualsClassName", "MaxLineLength")
    @Throws(IOException::class, InterruptedException::class)
    @JvmStatic
    fun main(args: Array<String>) {
        val parser = OptionParser()
        val input: OptionSpec<File?> =
            parser
                .accepts("input")
                .withRequiredArg()
                .ofType(File::class.java)
                .required()
                .describedAs(
                    "The name of the input file or directory. If a directory is given, all files ending in .cql will be processed"
                )
        val model: OptionSpec<File?> =
            parser
                .accepts("model")
                .withRequiredArg()
                .ofType(File::class.java)
                .describedAs(
                    "The name of an input file containing the model info to use for translation. Model info can also be provided through an implementation of ModelInfoProvider"
                )
        val output: OptionSpec<File?> =
            parser
                .accepts("output")
                .withRequiredArg()
                .ofType(File::class.java)
                .describedAs(
                    "The name of the output file or directory. If no output is given, an output file name is constructed based on the input name and target format"
                )
        val format: OptionSpec<CqlTranslator.Format> =
            parser
                .accepts("format")
                .withRequiredArg()
                .ofType(CqlTranslator.Format::class.java)
                .defaultsTo(CqlTranslator.Format.XML)
                .describedAs("The target format for the output")
        val rootDir: OptionSpec<File?> =
            parser
                .accepts("root-dir")
                .withOptionalArg()
                .ofType(File::class.java)
                .describedAs("Root directory of a FHIR IG project, used to resolve CQL namespaces")
        val disableDefaultModelInfoLoad: OptionSpec<*>? =
            parser.accepts("disable-default-modelinfo-load")
        val verify: OptionSpec<*>? = parser.accepts("verify")
        val optimization: OptionSpec<*>? = parser.accepts("date-range-optimization")
        val annotations: OptionSpec<*>? = parser.accepts("annotations")
        val locators: OptionSpec<*>? = parser.accepts("locators")
        val resultTypes: OptionSpec<*>? = parser.accepts("result-types")
        val detailedErrors: OptionSpec<*>? = parser.accepts("detailed-errors")
        val errorLevel: OptionSpec<*> =
            parser
                .accepts("error-level")
                .withRequiredArg()
                .ofType(CqlCompilerException.ErrorSeverity::class.java)
                .defaultsTo(CqlCompilerException.ErrorSeverity.Info)
                .describedAs(
                    "Indicates the minimum severity message that will be reported. If no error-level is specified, all messages will be output"
                )
        val disableListTraversal: OptionSpec<*>? = parser.accepts("disable-list-traversal")
        val disableListDemotion: OptionSpec<*>? = parser.accepts("disable-list-demotion")
        val disableListPromotion: OptionSpec<*>? = parser.accepts("disable-list-promotion")
        val enableIntervalDemotion: OptionSpec<*>? = parser.accepts("enable-interval-demotion")
        val enableIntervalPromotion: OptionSpec<*>? = parser.accepts("enable-interval-promotion")
        val disableMethodInvocation: OptionSpec<*>? = parser.accepts("disable-method-invocation")
        val requireFromKeyword: OptionSpec<*>? = parser.accepts("require-from-keyword")
        val strict: OptionSpec<*>? = parser.accepts("strict")
        val debug: OptionSpec<*>? = parser.accepts("debug")
        val validateUnits: OptionSpec<*>? = parser.accepts("validate-units")
        val signatures: OptionSpec<SignatureLevel> =
            parser
                .accepts("signatures")
                .withRequiredArg()
                .ofType(SignatureLevel::class.java)
                .defaultsTo(SignatureLevel.None)
                .describedAs(
                    "Indicates whether signatures should be included for invocations in the output ELM. Differing will include invocation signatures that differ from the declared signature. Overloads will include declaration signatures when the operator or function has more than one overload with the same number of arguments as the invocation"
                )
        val compatibilityLevel: OptionSpec<String?> =
            parser
                .accepts("compatibility-level")
                .withRequiredArg()
                .ofType(String::class.java)
                .describedAs(
                    "Compatibility level for the translator, valid values are 1.3, 1.4, and 1.5"
                )

        @Suppress("SpreadOperator") val options = parser.parse(*args)

        val source = input.value(options)!!.toPath()
        val destination =
            if (output.value(options) != null) output.value(options)!!.toPath()
            else if (source.toFile().isDirectory()) source else source.parent
        val outputFormat = format.value(options)
        val signatureLevel = signatures.value(options)

        val inOutMap: MutableMap<Path?, Path?> = HashMap()
        if (source.toFile().isDirectory()) {
            require(!(destination.toFile().exists() && !destination.toFile().isDirectory())) {
                "Output must be a valid folder if input is a folder!"
            }

            Files.walkFileTree(
                source,
                object : SimpleFileVisitor<Path>() {
                    @Throws(IOException::class)
                    override fun visitFile(
                        file: Path,
                        attrs: BasicFileAttributes,
                    ): FileVisitResult {
                        if (
                            file.toFile().getName().endsWith(".cql") ||
                                file.toFile().getName().endsWith(".CQL")
                        ) {
                            val destinationFolder =
                                destination.resolve(source.relativize(file.parent))
                            if (
                                !destinationFolder.toFile().exists() &&
                                    !destinationFolder.toFile().mkdirs()
                            ) {
                                System.err.printf("Problem creating %s%n", destinationFolder)
                            }
                            inOutMap[file] = destinationFolder
                        }
                        return FileVisitResult.CONTINUE
                    }
                },
            )
        } else {
            inOutMap[source] = destination
        }

        for (inOut in inOutMap.entries) {
            val `in`: Path = inOut.key!!
            var out: Path = inOut.value!!
            if (out.toFile().isDirectory()) {
                // Use input filename with ".xml", ".json", or ".coffee" extension
                var name = `in`.toFile().getName()
                if (name.lastIndexOf('.') != -1) {
                    name = name.take(name.lastIndexOf('.'))
                }
                name +=
                    when (outputFormat) {
                        CqlTranslator.Format.JSON -> ".json"
                        CqlTranslator.Format.COFFEE -> ".coffee"
                        CqlTranslator.Format.XML -> ".xml"
                        else -> ".xml"
                    }
                out = out.resolve(name)
            }

            require(out != `in`) { "input and output file must be different!" }

            var igContext: IGContext? = null
            if (options.has(rootDir)) {
                igContext = IGContext()
                igContext.initializeFromIni(
                    options.valueOf("root-dir").toString() + File.separator + "ig.ini"
                )
            }

            var modelProvider: ModelInfoProvider? = null
            if (options.has(model)) {
                val modelFile = options.valueOf<File>(model)
                require(!(!modelFile.exists() || modelFile.isDirectory())) {
                    "model must be a valid file!"
                }
                modelProvider = getModelInfoProvider(modelFile)
            }

            writeELM(
                `in`,
                out,
                outputFormat,
                modelProvider,
                igContext,
                CqlCompilerOptions(
                    options.has(optimization),
                    options.has(debug) || options.has(annotations),
                    options.has(debug) || options.has(locators),
                    options.has(debug) || options.has(resultTypes),
                    options.has(verify),
                    options.has(detailedErrors), // Didn't include in debug, maybe should...
                    if (options.has(errorLevel))
                        options.valueOf(errorLevel) as CqlCompilerException.ErrorSeverity?
                    else CqlCompilerException.ErrorSeverity.Info,
                    options.has(strict) || options.has(disableListTraversal),
                    options.has(strict) || options.has(disableListDemotion),
                    options.has(strict) || options.has(disableListPromotion),
                    options.has(enableIntervalDemotion),
                    options.has(enableIntervalPromotion),
                    options.has(strict) || options.has(disableMethodInvocation),
                    options.has(requireFromKeyword),
                    options.has(validateUnits),
                    options.has(disableDefaultModelInfoLoad),
                    signatureLevel,
                    if (options.has(compatibilityLevel)) options.valueOf(compatibilityLevel)!!
                    else "1.5",
                ),
            )
        }
    }
}
